pilot/pkg/networking/core/v1alpha3/cluster.go

```diff
@@ -931,6 +930,10 @@ func applyUpstreamTLSSettings(opts *buildClusterOpts, tls *networking.ClientTLSS
 				},
 				defaultTransportSocketMatch,
 			}
+		} else {
```

Comment:
Instead of resetting this here - What do you think about deriving appropriate subset traffic policy by merging destination and subset (for the values overridden) and applyTrafficPolicy only once. That avoid this special handling and would be consistent IMO.

Add a method like 
subsetTrafficPolicy(destinationRule.TrafficPolicy) {
trafficPolicy := &v1alpha3.TrafficPolicy{}
			if destinationRule.TrafficPolicy != nil {
				trafficPolicy.OutlierDetection = destinationRule.TrafficPolicy.OutlierDetection
				trafficPolicy.ConnectionPool = destinationRule.TrafficPolicy.ConnectionPool
				trafficPolicy.LoadBalancer = destinationRule.TrafficPolicy.LoadBalancer
				trafficPolicy.PortLevelSettings = destinationRule.TrafficPolicy.PortLevelSettings
				trafficPolicy.Tls = destinationRule.TrafficPolicy.Tls
			}

			// Override with subset values.
			if subset.TrafficPolicy.ConnectionPool != nil {
				trafficPolicy.ConnectionPool = subset.TrafficPolicy.ConnectionPool
			}
			if subset.TrafficPolicy.OutlierDetection != nil {
				trafficPolicy.OutlierDetection = subset.TrafficPolicy.OutlierDetection
			}
			if subset.TrafficPolicy.LoadBalancer != nil {
				trafficPolicy.LoadBalancer = subset.TrafficPolicy.LoadBalancer
			}
			if subset.TrafficPolicy.PortLevelSettings != nil {
				trafficPolicy.PortLevelSettings = subset.TrafficPolicy.PortLevelSettings
			}
			if subset.TrafficPolicy.Tls != nil {
				trafficPolicy.Tls = subset.TrafficPolicy.Tls
			}
```
And applyTrafficPolicy only once for subset?

Reply:
@ramaraochavali I am fine with investigating a refactor of the subset policy enforcement - but I suggest doing it outside of this PR.  There is an upcoming release and this fixes a longstanding bug in the istio.io egress gateway tasks.

Reply:
Sure. I am fine with that.

Reply:
that makes sense to me. For peer authentication policy we follow similiar manner, policy from root namespace and each namespace and workload level are first combined together in one place, and the config translation is another part.

Though this is more of networking's own code organization. Just commenting from an outsider perspective.